src/Data/Vector/Mutable/Linear.hs

```diff
@@ -58,93 +61,105 @@ import GHC.Exts hiding (fromList)
 
--- | Allocate a constant vector of a given non-zero size (and error on a bad
+empty :: (Vector a #-> Unrestricted b) -> Unrestricted b
+empty f = Array.empty (\arr -> f (Vec 0 arr))
+
+-- | Allocate a constant vector of a given non-negative size (and error on a bad
 -- size)
-constant :: HasCallStack =>
-  Int -> a -> (Vector a #-> Unrestricted b) -> Unrestricted b
+constant :: Int -> a -> (Vector a #-> Unrestricted b) -> Unrestricted b
 constant size x f
-  | size <= 0 = error ("Trying to construct a vector of size " ++ show size)
-  | otherwise = f $ Vec (size, size) (Unsafe.newMutArr size x)
+  | size < 0 = error ("Trying to construct a vector of size " ++ show size)
+  | otherwise = fromList (replicate size x) f
 
--- XXX: long line below
--- | Allocator from a non-empty list (and error on empty lists)
+-- | Creates a vector from a list.
 fromList :: HasCallStack => [a] -> (Vector a #-> Unrestricted b) -> Unrestricted b
-fromList [] _ = error "Trying to allocate a vector from an empty list"
-fromList xs@(x:_) (f :: Vector a #-> Unrestricted b) =
-  constant (Prelude.length xs) x buildThenRun
-  where
-    buildThenRun :: Vector a #-> Unrestricted b
-    buildThenRun vec = f $ doWrites (zip xs [0..]) vec
-
-    doWrites :: [(a,Int)] -> Vector a #-> Vector a
-    doWrites [] vec = vec
-    doWrites ((a,ix):ws) vec = doWrites ws (write vec ix a)
+fromList xs f =
+  Array.fromList xs Prelude.$ \arr ->
+    Array.size arr & \(arr', Unrestricted i) ->
+      f (Vec i arr')
 
 -- | Length of the vector
-length :: Vector a #-> (Vector a, Int)
-length = Unsafe.toLinear unsafeLength
-  where
-    unsafeLength :: Vector a -> (Vector a, Int)
-    unsafeLength v@(Vec (len, _) _) = (v, len)
+length :: Vector a #-> (Vector a, Unrestricted Int)
+length (Vec len arr) =
+  (Vec len arr, Unrestricted len)
+
+-- | Capacity of the vector
+capacity :: Vector a #-> (Vector a, Unrestricted Int)
+capacity (Vec len arr) =
+  Array.size arr & \(arr', cap) ->
+    (Vec len arr', cap)
 
 -- | Insert at the end of the vector
-snoc :: HasCallStack => Vector a #-> a -> Vector a
-snoc (Vec (len, size) ma) x
-  | len < size = write (Vec (len + 1, size) ma) len x
-  | otherwise = write (unsafeResize (size * 2) (Vec (len + 1, size) ma)) len x
-
--- | Write to an element already written to before.  Note: this will not write
--- to elements beyond the current length of the array and will error instead.
-write :: HasCallStack => Vector a #-> Int -> a -> Vector a
-write = Unsafe.toLinear writeUnsafe
-  where
-    writeUnsafe :: Vector a -> Int -> a -> Vector a
-    writeUnsafe v@(Vec (len, _) mutArr) ix val
-      | indexInRange len ix = case Unsafe.writeMutArr mutArr ix val of () -> v
-      | otherwise = error "Write index not in range."
-
--- | Read from a vector, with an in-range index and error for an index that is
--- out of range (with the usual range @0..length-1@).
-read :: HasCallStack => Vector a #-> Int -> (Vector a, a)
```

Comment:
The new version of this function returns a `Unrestricted a` instead. Since the vector does not contain its values linearly, I think it was overly restrictive.